.github/azure-pipelines.yaml

```diff
@@ -0,0 +1,31 @@
+trigger:
+- master
+
+pr:
+- master
+
+resources:
+  repositories:
+  - repository: azure_template
+    type: github
+    name: mdolab/.github
+    endpoint: mdolab
+
+pool:
+    vmImage: 'ubuntu-latest'
+
+extends:
+  template: azure/azure_template.yaml@azure_template
+  parameters:
+    REPO_NAME: adflow
+    IGNORE_STYLE: true
+    COMPLEX: true
+    GCC_CONFIG: config/defaults/config.LINUX_GFORTRAN.mk
+    INTEL_CONFIG: config/defaults/config.LINUX_INTEL.mk
+    INPUT_FILES: true
+    STEPS_MAKE_REAL: make
```

Comment:
The `STEPS_MAKE` and `STEPS_INSTALL` params are the default values right? So they don't necessarily need to be specified here.

Reply:
This depends on how we want to handle repos that are pure python. If we want a conditional that will skip any steps to copy config files and make fortran if the make steps are empty, then we need to specify the lines in this file. Otherwise, we can use an additional boolean and have make defaults.

Reply:
The approach I suggested was similar to 2, but instead of relying on `GCC_CONFIG` etc. to determine whether we have a build step, to check if `Makefile` exists. Similarly check `Makefile_CS` for whether complex build exists. This would be more general (since repos like TACS do not have these config files but still have make steps).

For these "weird" repos I also suggested having a bash script (stored alongside the yaml) for how to build the repo, instead of throwing the bash commands into this yaml file and cluttering it.

Reply:
Following Neil’s recommendation, I implemented the check for make files with the following lines:

```
  - script: |
      if { [[ ${{ parameters.BUILD_TYPE }} == 'real' ]] && [ -f "Makefile" ]; } || { [[ ${{ parameters.BUILD_TYPE }} == 'complex' ]] && [ -f "Makefile_CS" ]; }; then
        docker exec app /bin/bash -c ". \$HOME/.bashrc_mdolab && cd ${{ variables.DOCKER_WORKING_DIR }} && cp $(CONFIG_FILE) config/config.mk";
        docker exec app /bin/bash -c ". \$HOME/.bashrc_mdolab && cd ${{ variables.DOCKER_WORKING_DIR }} && ${{ parameters.STEPS_MAKE }}";
      fi
    displayName: Make

```

This is in the template (`.github` repo) so the change will take effect behind the scenes for this PR. `parameters.BUILD_TYPE` is hidden to the user and managed internally in the templates.

In terms of bash scripting make commands, I tested that already and it works fine, you just put the path of the bash script instead of the command in the respective variables. That being said, we may need to adjust the template if there is no config file to copy over.

Reply:
Is the `STEPS_MAKE` param set depending on `BUILD_TYPE`? For the config file, we can make it so that if it's an empty string, don't do the copying.

Reply:
No, build type is set as an additional constant that essentially just provides a flag. This ensures each test looks for the correct makefile:

```
- stage:
  dependsOn: []
  displayName: Test Real
  jobs:
  - template: azure_build.yaml
    parameters:
      REPO_NAME: ${{ parameters.REPO_NAME }}
      GCC_CONFIG: ${{ parameters.GCC_CONFIG }}
      INTEL_CONFIG: ${{ parameters.INTEL_CONFIG }}
      INPUT_FILES: ${{ parameters.INPUT_FILES }}
      STEPS_MAKE: ${{ parameters.STEPS_MAKE_REAL }}
      STEPS_INSTALL: ${{ parameters.STEPS_INSTALL_REAL }}
      STEPS_TEST: ${{ parameters.STEPS_TEST_REAL }}
      BUILD_TYPE: real

- stage:
  dependsOn: []
  displayName: Test Complex
  condition: eq(${{ parameters.COMPLEX }}, true)
  jobs:
  - template: azure_build.yaml
    parameters:
      REPO_NAME: ${{ parameters.REPO_NAME }}
      GCC_CONFIG: ${{ parameters.GCC_CONFIG }}
      INTEL_CONFIG: ${{ parameters.INTEL_CONFIG }}
      INPUT_FILES: ${{ parameters.INPUT_FILES }}
      STEPS_MAKE: ${{ parameters.STEPS_MAKE_COMPLEX }}
      STEPS_INSTALL: ${{ parameters.STEPS_INSTALL_COMPLEX }}
      STEPS_TEST: ${{ parameters.STEPS_TEST_COMPLEX }}
      BUILD_TYPE: complex
```

For the config file, we can implement that check as well.

Reply:
Okay that makes sense, I wanted to check that `STEPS_MAKE` was different dependent on whether it's a real or complex build.